exchanges/deribit/deribit_websocket.go

```diff
@@ -0,0 +1,1256 @@
+	rawUsersOrdersWithKindCurrencyAndIntervalChannel = "user.orders"
+	userPortfolioChannel                             = "user.portfolio"
+	userTradesChannelByInstrument                    = "user.trades."
+	userTradesByKindCurrencyAndIntervalChannel       = "user.trades"
+)
+
+var (
+	defaultSubscriptions = []string{
+		chartTradesChannel, // chart trades channel to fetch candlestick data.
+		orderbookChannel,
+		tickerChannel,
+		tradesWithKindChannel,
+	}
+
+	indexENUMS = []string{"ada_usd", "algo_usd", "avax_usd", "bch_usd", "bnb_usd", "btc_usd", "doge_usd", "dot_usd", "eth_usd", "link_usd", "ltc_usd", "luna_usd", "matic_usd", "near_usd", "shib_usd", "sol_usd", "trx_usd", "uni_usd", "usdc_usd", "xrp_usd", "ada_usdc", "bch_usdc", "algo_usdc", "avax_usdc", "btc_usdc", "doge_usdc", "dot_usdc", "bch_usdc", "bnb_usdc", "eth_usdc", "link_usdc", "ltc_usdc", "luna_usdc", "matic_usdc", "near_usdc", "shib_usdc", "sol_usdc", "trx_usdc", "uni_usdc", "xrp_usdc", "btcdvol_usdc", "ethdvol_usdc"}
+
+	pingMessage = WsSubscriptionInput{
+		ID:             2,
+		JSONRPCVersion: rpcVersion,
+		Method:         "public/test",
+		Params:         map[string][]string{},
+	}
+	setHeartBeatMessage = wsInput{
+		ID:             1,
+		JSONRPCVersion: rpcVersion,
+		Method:         "public/set_heartbeat",
+		Params: map[string]interface{}{
+			"interval": 15,
+		},
+	}
+)
+
+// WsConnect starts a new connection with the websocket API
+func (d *Deribit) WsConnect() error {
+	if !d.Websocket.IsEnabled() || !d.IsEnabled() {
+		return stream.ErrWebsocketNotEnabled
+	}
+	var dialer websocket.Dialer
+	err := d.Websocket.Conn.Dial(&dialer, http.Header{})
+	if err != nil {
+		return err
+	}
+	d.Websocket.Wg.Add(1)
+	go d.wsReadData()
+	if d.Websocket.CanUseAuthenticatedEndpoints() {
+		err = d.wsLogin(context.TODO())
+		if err != nil {
+			log.Errorf(log.ExchangeSys, "%v - authentication failed: %v\n", d.Name, err)
+			d.Websocket.SetCanUseAuthenticatedEndpoints(false)
+		}
+	}
+	return d.Websocket.Conn.SendJSONMessage(setHeartBeatMessage)
+}
+
+func (d *Deribit) wsLogin(ctx context.Context) error {
+	if !d.IsWebsocketAuthenticationSupported() {
+		return fmt.Errorf("%v AuthenticatedWebsocketAPISupport not enabled", d.Name)
+	}
+	creds, err := d.GetCredentials(ctx)
+	if err != nil {
+		return err
+	}
+	d.Websocket.SetCanUseAuthenticatedEndpoints(true)
+	n := d.Requester.GetNonce(nonce.UnixNano).String()
+	strTS := strconv.FormatInt(time.Now().UnixMilli(), 10)
+	str2Sign := strTS + "\n" + n + "\n"
+	hmac, err := crypto.GetHMAC(crypto.HashSHA256,
+		[]byte(str2Sign),
+		[]byte(creds.Secret))
+	if err != nil {
+		return err
+	}
+
+	request := wsInput{
+		JSONRPCVersion: rpcVersion,
+		Method:         "public/auth",
+		ID:             d.Websocket.Conn.GenerateMessageID(false),
+		Params: map[string]interface{}{
+			"grant_type": "client_signature",
+			"client_id":  creds.Key,
+			"timestamp":  strTS,
+			"nonce":      n,
+			"signature":  crypto.HexEncodeToString(hmac),
+		},
+	}
```

Comment:
This isn't working with testnet:
`"testnet":false`
full response:
```
{"jsonrpc":"2.0","id":173407906,"error":{"message":"invalid_credentials","code":13004},"usIn":1715042216786000,"usOut":1715042216786248,"usDiff":248,"testnet":false}
```

I have testnet set to `true` in `_test` and REST functions work properly. Could you please look into that?

Reply:
It was not setting the WebSocket URL to the test one. I have updated this